swarm/network/simulations/overlay_test.go

```diff
@@ -0,0 +1,198 @@
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+)
+
+var (
+	nodeCount = 16
+)
+
+//This test is used to test the overlay simulation.
+//As the simulation is executed via a main, it is easily missed on changes
+//An automated test will prevent that
+//The test just connects to the simulations, starts the network,
+//starts the mocker, gets the number of nodes, and stops it again.
+//It also provides a documentation on the steps needed by frontends
+//to use the simulations
+func TestOverlaySim(t *testing.T) {
+	//start the simulation
+	log.Info("Start simulation backend")
+	go runOverlaySim()
+
+	//the simulation runs usually at localhost:8888
+	localhostAddr := fmt.Sprintf("http://localhost:%d", httpSimPort)
+	//wait for the network to start
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	log.Debug("Waiting for http simulation server to be up...")
+	pingServer := func(ctx context.Context) {
+		ticker := time.NewTicker(50 * time.Millisecond)
+		for {
+			select {
+			case <-ticker.C:
+				resp, err := http.Get(localhostAddr)
+				if err == nil && resp.StatusCode == http.StatusOK {
+					return
+				}
+			case <-ctx.Done():
+				t.Fatal(errors.New("Timeout pinging http server"))
+			}
+		}
+	}
+	pingServer(ctx)
+
+	log.Debug("Http simulation server started. Start simulation network")
+	//start the simulation network (initialization of simulation)
+	resp, err := http.Post(localhostAddr+"/start", "application/json", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("Expected Status Code %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+	resp.Body.Close()
+
+	log.Debug("Start mocker")
+	//start the mocker, needs a node count and an ID
+	formstr := []byte(fmt.Sprintf("node-count=%d&mocker-type=%s", nodeCount, simulations.GetMockerList()[0]))
+
+	resp, err = http.Post(localhostAddr+"/mocker/start", "application/x-www-form-urlencoded", bytes.NewReader(formstr))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		reason, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+		log.Error(string(reason))
+		resp.Body.Close()
+		t.Fatalf("Expected Status Code %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	//variables needed to wait for all subscriptions established before uploading
+	var upCount int
+	fmt.Println("33")
+	trigger := make(chan discover.NodeID)
+	// quitC stops the event listener loops
+	// inside the step action method after step is complete
+	quitC := make(chan struct{})
+	defer close(quitC)
+
+	go watchSimEvents(ctx, trigger, quitC)
+
+	for {
+		<-trigger
+		//`nil` received, decrement count
+		upCount++
+		//all subscriptions received
+		if upCount == nodeCount {
+			break
+		}
+	}
+
+	log.Info("Get number of nodes")
+	//get the number of nodes
+	resp, err = http.Get(localhostAddr + "/nodes")
```

Comment:
It's probably worth using [the simulation client](https://godoc.org/github.com/ethereum/go-ethereum/p2p/simulations#Client) rather than making HTTP requests.

Reply:
Indeed. I'd like to keep this as-is though so to have more intuitive documentation for how to use the backend from a frontend which would issue HTTP requests.

Reply:
You mean documentation a bit like [this](https://github.com/ethereum/go-ethereum/tree/master/p2p/simulations#http-api)? :wink: